File: boilerplate/java/order-service/src/main/java/com/example/orderservice/application/usecases/RegisterUserUseCaseImpl.java
package com.example.orderservice.application.usecases;

import com.example.orderservice.application.dtos.RegisterCommand;
import com.example.orderservice.application.dtos.RegisterResult;
import com.example.orderservice.domain.events.UserRegistered;
import com.example.orderservice.domain.models.*;
import com.example.orderservice.domain.ports.EventPublisher;
import com.example.orderservice.domain.ports.PasswordHasher;
import com.example.orderservice.domain.ports.UserRepository;

import java.util.Set;

public class RegisterUserUseCaseImpl implements RegisterUserUseCase {

    private final UserRepository userRepository;
    private final PasswordHasher passwordHasher;
    private final EventPublisher eventPublisher;

    public RegisterUserUseCaseImpl(
            UserRepository userRepository,
            PasswordHasher passwordHasher,
            EventPublisher eventPublisher) {
        this.userRepository = userRepository;
        this.passwordHasher = passwordHasher;
        this.eventPublisher = eventPublisher;
    }

    @Override
    public RegisterResult execute(RegisterCommand command) {
        if (command == null) {
            throw new IllegalArgumentException("command must not be null");
        }

        Email email = new Email(command.email());

        if (userRepository.existsByEmail(email)) {
            throw new AuthenticationException("AUTH_EMAIL_EXISTS",
                "An account with this email already exists");
        }

        Password.validatePlaintext(command.password());
        String hashed = passwordHasher.hash(command.password());
        Password password = new Password(hashed);

        Set<Role> roles = command.roles() != null && !command.roles().isEmpty()
            ? Set.copyOf(command.roles())
            : Set.of(Role.USER);

        User user = User.create(email, password, roles);
        User saved = userRepository.save(user);

        eventPublisher.publish(new UserRegistered(saved.getId(), saved.getEmail().getValue()));

        return new RegisterResult(
            saved.getId().getValue().toString(),
            saved.getEmail().getValue(),
            saved.getRoles()
        );
    }
}

File: boilerplate/nestjs/order-service/src/application/usecases/register-user.use-case.impl.ts
import { Injectable, Inject } from "@nestjs/common";

import { IRegisterUserUseCase } from "./register-user.use-case.interface";
import { AuthenticationException } from "../../domain/exceptions/auth.exception";
import { Email } from "../../domain/models/email.value-object";
import { Password } from "../../domain/models/password.value-object";
import { Role } from "../../domain/models/role";
import { User } from "../../domain/models/user.aggregate";
import { IEventPublisher } from "../../domain/ports/event-publisher.port";
import { IPasswordHasher } from "../../domain/ports/password-hasher.port";
import { IUserRepository } from "../../domain/ports/user-repository.port";
import { RegisterCommand, RegisterResult } from "../dtos/user-order.dto";

@Injectable()
export class RegisterUserUseCaseImpl implements IRegisterUserUseCase {
  constructor(
    @Inject("IUserRepository")
    private readonly userRepository: IUserRepository,
    @Inject("IPasswordHasher")
    private readonly passwordHasher: IPasswordHasher,
    @Inject("IEventPublisher")
    private readonly eventPublisher: IEventPublisher,
  ) {}

  async execute(command: RegisterCommand): Promise<RegisterResult> {
    if (!command) {
      throw new Error("Command must not be null");
    }

    const email = new Email(command.email);

    if (await this.userRepository.existsByEmail(email)) {
      throw new AuthenticationException(
        "AUTH_EMAIL_EXISTS",
        "An account with this email already exists",
      );
    }

    const hashed = await this.passwordHasher.hash(command.password);
    const password = new Password(hashed);

    const roles =
      command.roles && command.roles.length > 0
        ? new Set(command.roles)
        : new Set([Role.USER]);

    const user = User.create(email, password, roles);
    const saved = await this.userRepository.save(user);

    await this.eventPublisher.publish({
      type: "UserRegistered",
      payload: {
        userId: saved.id.getValue(),
        email: saved.email.getValue(),
      },
    });

    return {
      userId: saved.id.getValue(),
      email: saved.email.getValue(),
      roles: saved.roles,
    };
  }
}
